balancer/rls/internal/cache.go

```diff
@@ -0,0 +1,403 @@
+	// will trigger a new RLS request.
+	timer *time.Timer
+	// callback provided by the LB policy to be notified when the backoff timer
+	// expires. This will trigger a new picker to be returned to gRPC, to force
+	// queued up RPCs to be retried.
+	callback func()
+}
+
+// lru is a cache implementation with a least recently used eviction policy.
+// Internally it uses a doubly linked list, with the least recently used element
+// at the front of the list and the most recently used element at the back of
+// the list. The value stored in this cache will be of type `cacheKey`.
+//
+// It is not safe for concurrent access.
+type lru struct {
+	ll *list.List
+
+	// A map from the value stored in the lru to its underlying list element is
+	// maintained to have a clean API. Without this, a subset of the lru's API
+	// would accept/return cacheKey while another subset would accept/return
+	// list elements.
+	m map[cacheKey]*list.Element
+}
+
+// newLRU creates a new cache with a least recently used eviction policy.
+func newLRU() *lru {
+	return &lru{
+		ll: list.New(),
+		m:  make(map[cacheKey]*list.Element),
+	}
+}
+
+func (l *lru) addEntry(key cacheKey) {
+	e := l.ll.PushBack(key)
+	l.m[key] = e
+}
+
+func (l *lru) makeRecent(key cacheKey) {
+	e := l.m[key]
+	l.ll.MoveToBack(e)
+}
+
+func (l *lru) removeEntry(key cacheKey) {
+	e := l.m[key]
+	l.ll.Remove(e)
+	delete(l.m, key)
+}
+
+func (l *lru) getLeastRecentlyUsed() cacheKey {
+	e := l.ll.Front()
+	if e == nil {
+		return cacheKey{}
+	}
+	return e.Value.(cacheKey)
+}
+
+func (l *lru) iterateAndRun(f func(cacheKey)) {
+	var next *list.Element
+	for e := l.ll.Front(); e != nil; e = next {
+		next = e.Next()
+		f(e.Value.(cacheKey))
+	}
+}
+
+// dataCache contains a cache of RLS data used by the LB policy to make routing
+// decisions.
+//
+// The dataCache will be keyed by the request's path and keys, represented by
+// the `cacheKey` type. It will maintain the cache keys in an `lru` and the
+// cache data, represented by the `cacheEntry` type, in a native map.
+//
+// It is not safe for concurrent access.
+type dataCache struct {
+	targetSize  int64 // Maximum allowed size.
+	currentSize int64 // Current size.
+	keys        *lru  // Cache keys maintained in lru order.
+	entries     map[cacheKey]*cacheEntry
+	logger      *internalgrpclog.PrefixLogger
+	shutdown    *grpcsync.Event
+}
+
+func newDataCache(size int64, logger *internalgrpclog.PrefixLogger) *dataCache {
+	return &dataCache{
+		targetSize: size,
+		keys:       newLRU(),
+		entries:    make(map[cacheKey]*cacheEntry),
+		logger:     logger,
+		shutdown:   grpcsync.NewEvent(),
+	}
+}
+
+// resize changes the maximum allowed size of the data cache.
+//
+// The return value indicates if an entry with a valid backoff timer was
+// evicted. This is important to the RLS LB policy which would send a new picker
+// on the channel to re-process any RPCs queued as a result of this backoff
+// timer.
+func (dc *dataCache) resize(size int64) bool {
```

Comment:
It might be a good idea to name the return value since it's not obvious (without reading the comment above) what it should mean.

Reply:
Done.

`updatePicker` was the best name i could think of for the return value. Let me know if you have a better one.

Reply:
I was thinking `backoffCanceled` like the existing variable name, but either way.

Reply:
Done.